File: LeetCode/799. 香槟塔.js
/**
 * @param {number} poured
 * @param {number} query_row
 * @param {number} query_glass
 * @return {number}
 */
var champagneTower = function (poured, query_row, query_glass) {
  let row = [poured]
  for (let i = 1; i <= query_row; i++) {
    const nextRow = Array.from({ length: i + 1 }, () => 0)
    for (let j = 0; j < i; j++) {
      const value = row[j]
      if (value > 1) {
        nextRow[j] += (value - 1) / 2
        nextRow[j + 1] += (value - 1) / 2
      }
    }
    row = nextRow
  }
  return Math.min(1, row[query_glass])
}

File: LeetCode/701-800/799. 香槟塔.c

#define MIN(a, b) ((a) < (b) ? (a) : (b))

double champagneTower(int poured, int query_row, int query_glass)
{
    double row[query_row + 1];
    row[0] = poured;
    for (int i = 1; i <= query_row; i++)
    {
        double nextRow[i + 1];
        for (int j = 0; j <= i; j++)
        {
            nextRow[j] = 0.0;
        }
        for (int j = 0; j < i; j++)
        {
            double volume = row[j];
            if (volume > 1)
            {
                nextRow[j] += (volume - 1) / 2;
                nextRow[j + 1] += (volume - 1) / 2;
            }
        }
        memcpy(row, nextRow, sizeof(double) * (i + 1));
    }
    return MIN(1.0, row[query_glass]);
}